Lib/test/test_unittest/test_runner.py

```diff
@@ -1452,6 +1452,205 @@ def test_1(self):
+        self.assertEqual(errored_test_obj.params.get("description"), "inner_subtest_expected_to_fail")
+
+        expected_events = [
+            "cm_body_fail_enter",
+            "before_body_raise",
+            ("cm_body_fail_exit", None),
+        ]
+        self.assertEqual(self.events, expected_events)
+
+    def test_enterContext_manager_enter_fails(self):
+        class MyTests(unittest.TestCase):
+            def runTest(inner_self):
+                events = TestSubTestCleanups._get_events()
+                with inner_self.subTest() as sub:
+                    cm_enter_fail = SubTestFailingEnterCM(events)
+                    with inner_self.assertRaisesRegex(RuntimeError, "Simulated __enter__ failure"):
+                        sub.enterContext(cm_enter_fail)
+                    inner_self.assertEqual(len(sub._cleanups), 0)
+
+        test = MyTests()
+        result = unittest.TestResult()
+        test.run(result)
+
+        self.assertTrue(result.wasSuccessful())
+        self.assertEqual(self.events, ["cm_enter_fail_enter_attempt"])
+
+    def test_failing_cleanup_function_in_subtest(self):
+        def failing_cleanup_func():
+            TestSubTestCleanups._get_events().append("failing_cleanup_called")
+            raise ZeroDivisionError("cleanup boom")
+
+        class MyTests(unittest.TestCase):
+            def runTest(inner_self):
+                recorder = TestSubTestCleanups._active_instance._record_event
+                with inner_self.subTest() as sub:
+                    sub.addCleanup(recorder, "cleanup_after_fail_attempt")
+                    sub.addCleanup(failing_cleanup_func)
+                    sub.addCleanup(recorder, "cleanup_before_fail_attempt")
+
+        test = MyTests()
+        result = unittest.TestResult()
+        test.run(result)
+
+        self.assertFalse(result.wasSuccessful())
+        self.assertEqual(len(result.errors), 1)
+
+        expected_events_order = [
+            "cleanup_before_fail_attempt",
+            "failing_cleanup_called",
+            "cleanup_after_fail_attempt",
+        ]
+        self.assertEqual(self.events, expected_events_order)
+
+    def test_multiple_subtests_independent_cleanups(self):
+        class MyTests(unittest.TestCase):
+            def runTest(inner_self):
+                events = TestSubTestCleanups._get_events()
+                recorder = TestSubTestCleanups._active_instance._record_event
+                for i in range(2):
+                    with inner_self.subTest(i=i) as sub:
+                        cleanup_msg = f"cleanup_sub_{i}"
+                        sub.addCleanup(recorder, cleanup_msg)
+                        events.append(f"subtest_body_{i}")
+
+        MyTests().run()
+
+        expected_events = [
+            "subtest_body_0", "cleanup_sub_0",
+            "subtest_body_1", "cleanup_sub_1",
+        ]
+        self.assertEqual(self.events, expected_events)
+
+    def tearDown(self):
```

Comment:
Can you put the tearDown() after the setUp() instead? it'll be easier to maintain and read.